velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -80,15 +123,227 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+    auto channel = probeType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterProbeProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(probeType->nameOf(channelValue));
+      types.emplace_back(probeType->childAt(channelValue));
+      continue;
+    }
+    channel = buildType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterBuildProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(buildType->nameOf(channelValue));
+      types.emplace_back(buildType->childAt(channelValue));
+      continue;
+    }
+    VELOX_FAIL(
+        "Join filter field {} not in probe or build input, filter: {}",
+        field->toString(),
+        filter->toString());
+  }
+
+  filterInputType_ = ROW(std::move(names), std::move(types));
+}
+
+RowVectorPtr NestedLoopJoinProbe::getMismatchedOutput(
+    RowVectorPtr data,
+    const SelectivityVector& matched,
+    BufferPtr mapping,
+    const std::vector<IdentityProjection>& projections,
+    const std::vector<IdentityProjection>& nullProjections) {
+  if (matched.isAllSelected()) {
+    return nullptr;
+  }
+
+  auto rawMapping = initializeRowNumberMapping(mapping, data->size(), pool());
+  int32_t numUnmatched{0};
+  for (auto i = 0; i < data->size(); ++i) {
+    if (!matched.isValid(i)) {
+      rawMapping[numUnmatched++] = i;
+    }
+  }
+  VELOX_CHECK_GT(numUnmatched, 0);
+
+  auto output =
+      BaseVector::create<RowVector>(outputType_, numUnmatched, pool());
+  projectChildren(output, data, projections, numUnmatched, mapping);
+  for (auto projection : nullProjections) {
+    output->childAt(projection.outputChannel) = BaseVector::createNullConstant(
+        outputType_->childAt(projection.outputChannel), output->size(), pool());
+  }
+  return output;
+}
 
+void NestedLoopJoinProbe::finishProbeInput() {
+  VELOX_CHECK_NOT_NULL(input_);
+  input_.reset();
+  buildIndex_ = 0;
+  if (!noMoreInput_) {
+    return;
+  }
+  if (!needsBuildMismatch(joinType_)) {
+    setState(ProbeOperatorState::kFinish);
+    return;
+  }
+  beginBuildMismatch();
+}
+
+void NestedLoopJoinProbe::noMoreInput() {
+  Operator::noMoreInput();
+  if (state_ != ProbeOperatorState::kRunning || input_ != nullptr) {
+    return;
+  }
+  if (!needsBuildMismatch(joinType_)) {
+    setState(ProbeOperatorState::kFinish);
+    return;
+  }
+  beginBuildMismatch();
+}
+
+void NestedLoopJoinProbe::beginBuildMismatch() {
+  VELOX_CHECK(needsBuildMismatch(joinType_));
+
+  std::vector<ContinuePromise> promises;
+  std::vector<std::shared_ptr<Driver>> peers;
+  if (!operatorCtx_->task()->allPeersFinished(
+          planNodeId(), operatorCtx_->driver(), &future_, promises, peers)) {
+    VELOX_CHECK(future_.valid());
+    return;
+  }
+
+  lastProbe_ = true;
+  // From now on, buildIndex_ is used to indexing into buildMismatched_
+  buildIndex_ = 0;
```

Comment:
Can we have a check here that buildIndex_  is already zero as reset by the last processed input?

Reply:
Indeed, replaced assignment with VELOX_CHECK_EQ.